server/enclave/ecalls.cpp

```diff
@@ -106,30 +111,26 @@ void enclave_modelaggregator(unsigned char*** encrypted_accumulator,
             continue; // Didn't receive this variable from any clients
         }
 
-        vector<double> new_val(vars.size(), 0.0);
         double iters_sum = accumulate(n_local_iters.begin(), n_local_iters.end(), 0);
-        for (int i = 0; i < new_val.size(); i++) {
+        for (int i = 0; i < old_params[v_name].size(); i++) {
             for (vector<double> weights : vars) {
-                new_val[i] += weights[i];
+                old_params[v_name][i] += weights[i];
             }
-            new_val[i] /= iters_sum;
+            old_params[v_name][i] /= iters_sum;
         }
-        params[v_name] = new_val;
     }
 
-    string serialized_new_params = serialize(params);
+    string serialized_new_params = serialize(old_params);
 
-    unsigned char** encrypted_new_params = new unsigned char*[encryption_metadata_length * sizeof(char*)];
-    encrypt_bytes((unsigned char*) serialized_new_params.c_str(), serialized_new_params.length(), encrypted_new_params);
+    unsigned char** encrypted_new_params = new unsigned char*[encryption_metadata_length * sizeof(unsigned char*)];
+    encrypt_bytes((unsigned char*) serialized_new_params.c_str(), serialized_new_params.size(), encrypted_new_params);
 
-    // Need to copy over the encrypted model, IV, and tag over to untrusted memory
-    unsigned char** usr_addr_params = (unsigned char**) oe_host_malloc(encryption_metadata_length * sizeof(unsigned char *));
+    // Need to copy the encrypted model, IV, and tag over to untrusted memory
+    *encrypted_new_params_ptr = (unsigned char**) oe_host_malloc(encryption_metadata_length * sizeof(unsigned char*));
+    *new_params_length = serialized_new_params.size();
     for (int i = 0; i < encryption_metadata_length; i++) {
-        size_t item_length = strlen((const char *) encrypted_new_params[i]);
-        unsigned char* item = (unsigned char*) oe_host_malloc(item_length * sizeof(unsigned char));
-        memcpy((void*) item[i], encrypted_new_params[i], item_length * sizeof(unsigned char));
-        memcpy(usr_addr_params[i], item, sizeof(unsigned char *));
+        size_t item_length = i == 0 ? *new_params_length : strlen((const char *) encrypted_new_params[i]);
```

Comment:
The use of `strlen` here is incorrect: It will stop counting once it encounters a zero byte.